include/deal.II/numerics/vector_tools.templates.h

```diff
@@ -74,6 +77,81 @@
 DEAL_II_NAMESPACE_OPEN
 
 
+namespace
+{
+  template <int dim, int fe_degree, int components, typename number>
+  class MassOperator : public Subscriptor
```

Comment:
I think that the diagonal - or lumped diagonal, which is much easier to compute than the diagonal because you get it by applying the mass matrix on a vector of ones - would indeed be useful. As soon as the meshes are not exactly uniform or polynomial degrees are higher than 2, the different scaling indeed helps iteration counts.